String list (#269)
fortran-lang/stdlib: doc/specs/stdlib_stringlist.md

different lengths. Such collections are quite useful though and the
language allows us to define a derived type that can handle such
collections.

The `stdlib_stringlist` module defines a derived type that is capable of
storing a list of strings and of manipulating them.

Methods include:

* inserting strings at a given position
* replacing strings at a given position
* deleting a single string or a range of strings
* retrieving a string or a range of strings at a given position
* finding the position of a particular string or a string which contains some substring
* sorting the list

## Positions in a list of strings

The module implements what are effectively infinitely long lists: a position is
represented as a positive integer, but there is no "out-of-bound" index. That is,
the following piece of code will simply work:

```fortran
type(stringlist_type) :: list

! Add two strings ...
call list%insert( list_head, "The first string" )
call list%insert( 20, "The last string" )

write(*,*) 'The last:    ', list%get(list_end)
write(*,*) 'Beyond that: ', list%get(30)
```
The special position `list_head` represents the position *before* the first element. Likewise,

> the special position `list_end` represents the position of the *last* element. You can

Reviewer: This feels inconsistent to me. `list_head` represents an element outside of the list, while `list_end` represents an element in the list. I would prefer either having both special indices refer to elements inside or outside the list, but not mixed.

Reviewer: One surprising consequence of the current choice of `list_head` (`0`) and `list_end` (`-1`) is that appending with `insert` is impossible:
```f90
call list%insert(list_end + 1, "last item")
```
Because it will evaluate to `list_head` and prepend instead.

Reviewer: One surprising consequence of the current choice of `list_head` (`0`) and `list_end` (`-1`) is that appending with `insert` is impossible:
```f90
call list%insert(list_end + 1, "last item")
```
Because it will evaluate to `list_head` and prepend instead.

Author: Yes, there is that ... I hope to solve this with special types rather than special values.

use these positions to insert a string before the first string that is already in the
list or to insert after the last string that has been inserted.

If you specify a position beyond the last, the `list%get()` method simply returns an empty
string.

You can also specify *negative* positions, but they are interpreted as going back from the

> last inserted string. If you need the last but one string, you can do so innthis way:

Reviewer: ```suggestion
last inserted string. If you need the last but one string, you can do so in this way:
```

```fortran

> write(*,*) 'The last but onw: ', list%get(list_end-1)

Reviewer: ```suggestion
write(*,*) 'The last but one: ', list%get(list_end-1)
```

#### Status

Experimental

#### Description

Get the string at the given position.

#### Syntax

`string = list % [[stringlist_type(type):get(bound) ( idx )]]`

#### Class

Function

#### Arguments

`list`: the stringlist variable to retrieve a string from

`idx`: the index of the string to be retrieved (see [`the section on positions`](./stdlib_stringlist.html#position-in-a-list-of-strings)

#### Result value

A copy of the string stored at the indicated position.

### `index` - find the index of a particular string in the list

#### Status

Experimental

#### Description

Get the position of the first stored string that matches the given string, if `back` is not present or false. If `back` is
false, return the position of the last stored string that matches. Note that trailing blanks are ignored.

#### Syntax

`idx = list % [[stringlist_type(type):index(bound) ( string, back )]]`

#### Class

Function

#### Arguments

`list`: the stringlist variable to retrieve a string from

`string`: the string to be found in the list

`back` (optional): logical argument indicating the first occurrence should be returned (`false`) or the last (`true`)

#### Result value

> The result is either the index of the string in the list or -1 if the string was not found

Reviewer: Why is -1 chosen here? The intrinsics `index`/`scan`/`verify` return 0 on “failure” rather than -1.

Author: Yes, 0 is more consistent. Just corrected that

#### Example

Because trailing blanks are ignored, the following calls will give the same result:

```fortran
    write(*,*) list%index( 'A' )
    write(*,*) list%index( 'A    ' )
```

### `index_sub` - find the index of a string containing the given substring in the list

#### Status

Experimental

#### Description

Get the position of the first stored string that contains the given substring, if `back` is not present or false. If `back` is
false, return the position of the last stored string that contains it.

#### Syntax

`idx = list % [[stringlist_type(type):index_sub(bound) ( substring, back )]]`

#### Class

Function

#### Arguments

`list`: the stringlist variable to retrieve a string from

`substring`: the substring in question

`back` (optional): logical argument indicating the first occurrence should be returned (`false`) or the last (`true`)

#### Result value

> The result is either the index of the string in the list or -1 if the string was not found

Reviewer: Same question on the return code as with `index`.

Author: Indeed

### `sort` - return a sorted list

#### Status

Experimental

#### Description

Create a new list consisting of the sorted strings of the given list. The strings are sorted according to ASCII, either
in ascending order or descending order.

#### Syntax

`sortedlist = list % [[stringlist_type(type):sort(bound) ( ascending )]]`

#### Class

Subroutine

#### Arguments

`list`: the stringlist variable of which the contents should be copied

`ascending` (optional): if not present or true, sort the list in ascending order, otherwise descending

#### Result value

The contents of the given list is sorted and then stored in the new list.

### `=` - copy the contents of a list

#### Status

Experimental

#### Description

Copy an existing list to a new one. The original list remains unchanged.

#### Syntax

`copylist = list`

#### Class

Assignment

#### Operands

`list`: the stringlist variable to be copied

### `//` - concatenate a list with one or more strings

#### Status

Experimental

#### Description

Concatenate a list with a string, a list of strings or a plain array

#### Syntax

`concatenatedlist = list // string`

`concatenatedlist = string // list`

#### Class

> Assignment

Reviewer: ```suggestion
Operator
```

Author: The standard makes a difference between assignment and operator ... I can live with either, though.
